libcaf_io/src/default_multiplexer.cpp

```diff
@@ -990,341 +663,7 @@ int64_t default_multiplexer::next_endpoint_id() {
-  CAF_LOG_TRACE(CAF_ARG2("fd", fd()));
-  CAF_ASSERT(mgr != nullptr);
-  activate(mgr);
-}
-
-void datagram_handler::activate(datagram_manager* mgr) {
-  if (!reader_) {
-    reader_.reset(mgr);
-    event_handler::activate();
-    prepare_next_read();
-  }
-}
-
-void datagram_handler::ack_writes(bool x) {
-  ack_writes_ = x;
-}
-
-
-void datagram_handler::write(datagram_handle hdl, const void* buf,
-                             size_t num_bytes) {
-  wr_offline_buf_.emplace_back();
-  wr_offline_buf_.back().first = hdl;
-  auto cbuf = reinterpret_cast<const char*>(buf);
-  wr_offline_buf_.back().second.assign(cbuf,
-                                       cbuf + static_cast<ptrdiff_t>(num_bytes));
-}
-
-void datagram_handler::flush(const manager_ptr& mgr) {
-  CAF_ASSERT(mgr != nullptr);
-  CAF_LOG_TRACE(CAF_ARG(wr_offline_buf_.size()));
-  if (!wr_offline_buf_.empty() && !writing_) {
-    backend().add(operation::write, fd(), this);
-    writer_ = mgr;
-    writing_ = true;
-    prepare_next_write();
-  }
-}
-
-
-std::unordered_map<datagram_handle, ip_endpoint>& datagram_handler::endpoints() {
-  return ep_by_hdl_;
-}
-
-const std::unordered_map<datagram_handle, ip_endpoint>&
-datagram_handler::endpoints() const {
-  return ep_by_hdl_;
-}
-
-void datagram_handler::add_endpoint(datagram_handle hdl, const ip_endpoint& ep,
-                                    const manager_ptr mgr) {
-  auto itr = hdl_by_ep_.find(ep);
-  if (itr == hdl_by_ep_.end()) {
-    hdl_by_ep_[ep] = hdl;
-    ep_by_hdl_[hdl] = ep;
-    writer_ = mgr;
-  } else if (!writer_) {
-    writer_ = mgr;
-  } else {
-    CAF_LOG_ERROR("cannot assign a second servant to the endpoint "
-                  << to_string(ep));
-    abort();
-  }
-}
-
-void datagram_handler::remove_endpoint(datagram_handle hdl) {
-  CAF_LOG_TRACE(CAF_ARG(hdl));
-  auto itr = ep_by_hdl_.find(hdl);
-  if (itr != ep_by_hdl_.end()) {
-    hdl_by_ep_.erase(itr->second);
-    ep_by_hdl_.erase(itr);
-  }
-}
-
-void datagram_handler::stop_reading() {
-  CAF_LOG_TRACE("");
-  close_read_channel();
-  passivate();
-}
-
-void datagram_handler::removed_from_loop(operation op) {
-  switch (op) {
-    case operation::read: reader_.reset(); break;
-    case operation::write: writer_.reset(); break;
-    case operation::propagate_error: break;
-  };
-}
-
-void datagram_handler::prepare_next_read() {
-  CAF_LOG_TRACE(CAF_ARG(wr_buf_.second.size())
-                << CAF_ARG(wr_offline_buf_.size()));
-  rd_buf_.resize(max_datagram_size_);
-}
-
-void datagram_handler::prepare_next_write() {
-  CAF_LOG_TRACE(CAF_ARG(wr_offline_buf_.size()));
-  wr_buf_.second.clear();
-  if (wr_offline_buf_.empty()) {
-    writing_ = false;
-    backend().del(operation::write, fd(), this);
-  } else {
-    wr_buf_.swap(wr_offline_buf_.front());
-    wr_offline_buf_.pop_front();
-  }
-}
+// -- Related helper functions -------------------------------------------------
 
 class socket_guard {
```

Comment:
Since we're already touching this: this is a good opportunity to move the socket guard out of here. See https://github.com/actor-framework/actor-framework/issues/696.